### ISSUE TITLE
Support Async IEnumerable to IQueryable mapping extension (As(Async)Queryable)

### ISSUE THREAD

**Issue by swimmesberger** [dotnet/efcore#32893]
I encountered this issue while unit testing, to "mock" away a IQueryable<T> with hardcoded values. While investigating I found following things:

1. AsQueryable returns a System.Linq.EnumerableQuery (Runtime Library) instance that does not implement IAsyncQueryable
2. IAsyncQueryProvider is part of ef core so System.Linq.EnumerableQuery can't implement it
3. Ef Core could provide a AsAsyncQueryable extension method that returns a custom Microsoft.EntityFrameworkCore.Query.AsyncEnumerableQuery implementation
4. AsyncEnumerableQuery implements IAsyncQueryProvider and provides an implementation that simply maps async call to sync calls

This would allow usage patterns like the following:
```csharp
[Fact]
public async Task TestCountAsync() {
    var queryable = new[] {
        new WeatherForecastEntity {
            Id = new Guid("0224589c-bc34-4761-b8e6-8b95aba85a00"),
            Date = new DateOnly(2010, 01, 01),
            TemperatureC = 12,
            CreatedAt = new DateTimeOffset(2023, 10, 23, 0, 0, 0, TimeSpan.Zero)
        }
    }.AsAsyncQueryable();
    (await queryable.CountAsync()).Should().Be(1);
}
```

There is definitely a need when looking at following library:
https://www.nuget.org/packages/MockQueryable.EntityFrameworkCore
https://github.com/romantitov/MockQueryable
https://github.com/scott-xu/EntityFramework.Testing

The current Version has > 2.000.000 downloads.

Moving the code into ef core would ensure that it always stays up-to-date with the current ef implementation. Additionally (especially) for new users the 'IQueryable' doesn't implement 'IAsyncQueryProvider' exception is kind of unexpected when using AsQueryable, because most new code is async nowadays.

I quickly sketched something that wraps the System.Linq.EnumerableQuery implementation with additional async support:
https://github.com/swimmesberger/AsyncEnumerableQuery

**Comment by roji** [comment]
@swimmesberger the above is indeed possible (and your implementation looks good at first glance), but note that we discourage mocking IQueryables, as that is rare a path to success. Specifically, the point of IQueryable is the ability to compose additional operators on top, which are translated for evaluation e.g. in the database. When mocking IQueryable via a fake in-memory list, LINQ operators composed on top do not behave as they would if evaluated against the actual database, limiting the value of the test.

Our general recommendation is to instead run tests against your real database, rather than against a fake, typically by using something like [testcontainers.net](https://testcontainers.com/guides/getting-started-with-testcontainers-for-dotnet/) to easily and cheaply spin up a containerized database. Where people insist on testing against a fake, in-memory SQLite is a good alternative (that is closer to real SQL databases), and EF also has its in-memory provider for a true, in-memory fake. Please see our [testing docs](https://learn.microsoft.com/en-us/ef/core/testing/choosing-a-testing-strategy) for more info on all of this.

Specifically on your proposal, this is the first time AFAIK that someone has expressed the need for an AsAsyncQueryable() operator in EF, so it doesn't seem like many people are blocked by it... I think we'd probably want to see a bit more user interest before adding something like this.

**Comment by swimmesberger** [comment]
@roji Sure! Thanks for the detailed answer! 
I'm also a big advocate for using testcontainers in the integration testing layer of our testing pyramid. There we make sure that all ef core queries return the expected result from the database. 
In this case I'm on the unit testing layer where I want to test application logic. The Microsoft recommended approach is to abstract the data access layer with the [repository pattern](https://learn.microsoft.com/en-us/aspnet/mvc/overview/older-versions/getting-started-with-ef-5-using-mvc-4/implementing-the-repository-and-unit-of-work-patterns-in-an-asp-net-mvc-application#the-repository-and-unit-of-work-patterns). The repository pattern definitely helps here to make application logic more testable because repositories can be easily mocked supplying test data to the application logic. Unfortunately the repository pattern leads to more complicated code because you lose the flexibility of the IQueryable interface. Sure, one could also extract the core logic from the service in a separate class that abstracts away the data-source and simply takes (in-memory) data as input and returns the result, which makes the logic again more testable. This again introduces abstractions/complexity that might be overkill for certain applications.

With IQueryable I sought to find a good middle ground between data abstraction and flexibility for fast prototyping and unit-testing, that does not stand in my way while developing (repository) and still offers me the flexibility to provide in-memory data.

Using ef core in-memory/sqlite is (for me) too much overhead/complexity unnecessary for unit testing purposes. Integration tests as the second level of defense ensure proper execution with the concrete database implementation for the project (Testcontainers ❤) 

An example I have in mind is following:

```csharp
public interface IRepository<T> {
    IQueryable<T> Query();
    
    Task<T> AddAsync(T entity, CancellationToken cancellationToken = default);
    void Update(T entity);
    void Delete(T entity);
    
    Task<int> SaveChangesAsync(CancellationToken cancellationToken = default);
}
```

```csharp
public sealed class WeatherForecastService {
    private static Expression<Func<WeatherForecastEntity, WeatherForecastDto>> MapDto => 
        entity => new WeatherForecastDto(entity.Id, entity.Date, entity.TemperatureC, entity.Summary, entity.CreatedAt);
    private static Expression<Func<WeatherForecastEntity, object?>> DefaultOrder =>
        entity => entity.CreatedAt;
    
    private readonly IRepository<WeatherForecastEntity> _repository;
    
    public WeatherForecastService(IRepository<WeatherForecastEntity> repository) {
        _repository = repository;
    }
    
    public async Task<List<WeatherForecastDto>> GetForecasts(CancellationToken cancellationToken = default) {
        return await _repository.Query().Select(MapDto).ToListAsync(cancellationToken: cancellationToken);
    }

    public async Task<WeatherForecastDto?> GetForecastById(Guid id, CancellationToken cancellationToken = default) {
        return await _repository.Query()
            .Where(x => x.Id == id)
            .OrderBy(DefaultOrder)
            .Select(MapDto)
            .FirstOrDefaultAsync(cancellationToken);
    }
}
```

```csharp
public sealed class QueryMockRepository<T> : IRepository<T> {
    private readonly IReadOnlyList<T> _data;

    public QueryMockRepository(IReadOnlyList<T> data) {
        _data = data;
    }

    public IQueryable<T> Query() => _data.AsAsyncQueryable();

    public Task<T> AddAsync(T entity, CancellationToken cancellationToken = default) => throw new NotImplementedException();

    public void Update(T entity) => throw new NotImplementedException();

    public void Delete(T entity) => throw new NotImplementedException();

    public Task<int> SaveChangesAsync(CancellationToken cancellationToken = default) => throw new NotImplementedException();
}

public static class MockRepositoryExtensions {
    public static IRepository<T> AsRepository<T>(this IEnumerable<T> data) {
        return new QueryMockRepository<T>(data);
    }
}
```

```csharp
public class WeatherForecastServiceTest {
    private readonly ITestOutputHelper _testOutput;

    public WeatherForecastServiceTest(ITestOutputHelper testOutput) {
        _testOutput = testOutput;
    }

    [Fact]
    public async Task TestWeatherForecast() {
        var service = new WeatherForecastService(new WeatherForecastEntity[] {
            new() {
                Id = new Guid("0224589c-bc34-4761-b8e6-8b95aba85a00"),
                Date = new DateOnly(2010, 01, 01),
                TemperatureC = 12,
                CreatedAt = new DateTimeOffset(2023, 10, 23, 0, 0, 0, TimeSpan.Zero)
            }
        }.AsRepository());
        var forecasts = await service.GetForecasts();
        _testOutput.WriteLine(string.Join(',', forecasts));
        forecasts.Should().HaveCount(1);
    }
}
```

**Comment by roji** [comment]
> Unfortunately the repository pattern leads to more complicated code because you lose the flexibility of the IQueryable interface.

That's true, but that's exactly the point of separating your application code from data access code (in order to make your application testable). Any LINQ operator composed over data is effectively part of the data access code, and not your application; to fully understand this, consider that if you didn't have LINQ and were writing SQL directly, that operator would simply be part of your SQL statement. This is also why that LINQ operator really does work differently depending on the LINQ provider being used (and therefore varies behavior between in-memory fake databases and your real database).

So to actually separate data access code from application code, the query's LINQ operators must be themselves inside the repository abstraction, with the latter exposing only IEnumerable, not IQueryable. Otherwise you're leaking data access into your application code.

I get the desire for a "middle ground", but we generally feel like this approach is a real pit of failure that ends up causing more trouble than it helps. For true separation and unit testing, a real repository layer is necessary (so that entire queries can be abstracted away), and on the other end you can test against your actual production database via testcontainers.net.

> Using ef core in-memory/sqlite is (for me) too much overhead/complexity unnecessary for unit testing purposes. Integration tests as the second level of defense ensure proper execution with the concrete database implementation for the project (Testcontainers ❤)

I'm curious about this - compared to mocking IQueryable, what's the additional overhead/complexity of using the EF in-memory/SQLite providers? That's simply a matter of switching your provider, and you basically get a sort of mocked IQueryable that will likely be significantly more functional in any case (e.g. the SQLite provider even supports transactions, raw SQL, etc.). Note that this still isn't what I think people should do; but if you're absolutely looking to test against a database fake, and you absolutely want to avoid a repository, then it's not clear to my why you'd prefer mocking IQueryable rather than using the EF in-memory/SQLite providers.

**Comment by swimmesberger** [comment]
@roji Thank you for these detailed insights and opinions! 😊
What are you thoughts on the specification pattern as replacement for the "standard" repository pattern? As it is proposed by [Ardalis](https://github.com/ardalis/Specification/tree/main) .

>I'm curious about this - compared to mocking IQueryable, what's the additional overhead/complexity of using the EF in-memory/SQLite providers? That's simply a matter of switching your provider, and you basically get a sort of mocked IQueryable that will likely be significantly more functional in any case (e.g. the SQLite provider even supports transactions, raw SQL, etc.). Note that this still isn't what I think people should do; but if you're absolutely looking to test against a database fake, and you absolutely want to avoid a repository, then it's not clear to my why you'd prefer mocking IQueryable rather than using the EF in-memory/SQLite providers.

In-Memory/SQLite Provider requires DI to properly set things up because the dependency graph is pretty large. For Unit-Testing I want to reduce complexity as much as possible (no magic) so you can on first glance see what the code is doing.  I simply want to say "this is my data; do your thing; give me the result" without any indirections. Additionally I want my unit tests to be fast, to ensure I can execute them repeatedly as often as I want. 

All these things don't matter for my integration tests, here I can setup a container and do some DI setup without sacrificing DX.

**Comment by roji** [comment]
> What are you thoughts on the specification pattern as a replacement for the "standard" repository pattern? As it is proposed by [Ardalis](https://github.com/ardalis/Specification/tree/main).

I've taken a quick glance, and it seems to be a rather standard generic repository approach - or am I missing something? Note that DbSet itself already represents a sort of repository, in the sense that it contains APIs similar to the RepositoryBaseOfT in that repo (Add, FirstOrDefaultAsync...). The actual value of a repository start to come into play when they wrap complex queries inside them, exposing a simple, enumerable-returning method.

> In-Memory/SQLite Provider requires DI to properly set things up because the dependency graph is pretty large.

OK. Generally just setting up a DbContext with one of these fake providers seems simple enough (and lots of people do that - unfortunately), so I'd really recommend against mocking IQueryable. But then again, I'd recommend against using fake providers as well...

**Comment by roji** [comment]
> Unfortunately I don't like that it abstracts away the power of IQueryable.

So just to make sure this isn't missed - if you're looking for a "repository" exposing IQueryable, than EF's DbSet is already that. I can't really see a reason to wrap it inside yet another thing that exposes IQueryables. The point of a repository is to isolate the **entire** data access code behind a method, meaning that no queryable operators can remain outside (therefore the repository methods must return IEnumerable). Only that brings the actual benefits of a repository, i.e. the ability to completely mock out data access, the ability to swap database providers (where different LINQ queries may be necessary), etc.

**Comment by swimmesberger** [comment]
> > Unfortunately I don't like that it abstracts away the power of IQueryable.
> 
> So just to make sure this isn't missed - if you're looking for a "repository" exposing IQueryable, than EF's DbSet is already that. I can't really see a reason to wrap it inside yet another thing that exposes IQueryables. The point of a repository is to isolate the **entire** data access code behind a method, meaning that no queryable operators can remain outside (therefore the repository methods must return IEnumerable). Only that brings the actual benefits of a repository, i.e. the ability to completely mock out data access, the ability to swap database providers (where different LINQ queries may be necessary), etc.

Sure thing! I just wanted to note that it's a disadvantage without wishing to diminish the advantages.

> the ability to swap database providers (where different LINQ queries may be necessary), etc.
That's actually a non-goal for me. 

To summarize the ticket - I was just looking for a way to directly use IQueryable in application code without sacrificing testability. The easy way out was mocking IQueryable (AsyncEnumerable, InMemory Provider, SQLite InMemory, ...), which works! ...but might not work for all kind of query types or even behave differently. This can be mitigated by completely abstracting away the data access via IEnumerable types with the Repository pattern. This might slow down rapid development because you can't directly use the Ef core Repository DbSet<T> (IQueryable<T>) but offers better unit testability.

And again for anyone stumbling on this issue: Unit Tests are no replacement for proper database tests with something like https://testcontainers.com/ - I highly recommend that. My goal was business logic/service level tests.

**Comment by roji** [comment]
Thanks, yeah - I think we're generally aligned here.

To go back to the original question, I think it makes theoretical sense to introduce AsAsyncQueryable() into EF, mirroring the standard AsQueryable(). However, it seems that the only practical use for that would be for mocking IQueryable, which is a testing strategy we recommend against - so I'm not sure I see us actually doing this change.

We can also keep this in the backlog to see if other users request it.

**Comment by DunetsNM** [comment]
>   I think it makes theoretical sense to introduce AsAsyncQueryable() into EF, mirroring the standard AsQueryable(). However, it seems that the only practical use for that would be for mocking IQueryable

Let me disagree with this. I think that there's no need to have separate `AsAsyncQueryable()`  at all because it's implementation detail. Instead this should be default behavior of `AsQueryable` i.e. `enumerable.AsQueryable().ToListAsync()` should simply work, even if "async" enumeration means simply do it sync and pass to `Task.FromResult` under the hood.  I don't understand why is this not the default behavior. The whole point of LINQ is to have type-safe queries that don't fail at runtime with an arbitrary type coercion error. Current "if it compiles then it runs but not always" behavior is just deceptive.

There should be ways to achieve this without compromising performance of sync extension methods.  Does it just reduce to this piece of source code in `System.Data.Entity.QueryableExtensions`?

```
private static IDbAsyncEnumerable<T> AsDbAsyncEnumerable<T>(this IQueryable<T> source)
{
    return source is IDbAsyncEnumerable<T> dbAsyncEnumerable ? dbAsyncEnumerable : throw Error.IQueryable_Not_Async((object) ("<" + typeof (T)?.ToString() + ">"));
}
```

As far as I can see, it's called only when an `IQueryable<>` is enumerated via an async extension method (such as `ToListAsync`, `CountAsync` etc.)

So instead of throwing exception when `source` doesn't implement the `IDbAsyncEnumerable<>` we can just wrap `source` in a trivial adapter like this one

```
    private class DummyDbAsyncEnumerable<T> : IDbAsyncEnumerable<T>
    {
        private readonly IEnumerable<T> _enumerable;
        public DummyDbAsyncEnumerable(IEnumerable<T> enumerable) { _enumerable = enumerable; }
        public IDbAsyncEnumerator<T> GetAsyncEnumerator() => new EnumerableAsyncFriendlyEnumerator<T>(_enumerable.GetEnumerator());
        IDbAsyncEnumerator IDbAsyncEnumerable.GetAsyncEnumerator() => GetAsyncEnumerator();
    }

    private class EnumerableAsyncFriendlyEnumerator<T> : IDbAsyncEnumerator<T>
    {
        private readonly IEnumerator<T> _enumerator;
        public EnumerableAsyncFriendlyEnumerator(IEnumerator<T> enumerator)  {  _enumerator = enumerator;  }
        public void Dispose() => _enumerator.Dispose();
        public Task<bool> MoveNextAsync(CancellationToken cancellationToken) =>
            Task.FromResult(!cancellationToken.IsCancellationRequested && _enumerator.MoveNext());

        public T Current => _enumerator.Current;
        object IDbAsyncEnumerator.Current => Current;
    }

...

private static IDbAsyncEnumerable<T> AsDbAsyncEnumerable<T>(this IQueryable<T> source)
{
    return source is IDbAsyncEnumerable<T> dbAsyncEnumerable ? dbAsyncEnumerable : new DummyDbAsyncEnumerable<T>(source);
}
``` 

Do I miss anything here?

==

As for use cases beyond just testing.  I'm working with a legacy codebase that has a lot of blocking EF calls inherited from `IEnumerable<>` (e.g. `ToList` or `ToArray`) that should be changed to async equivalents.  There's a lot of code and I want to simply blanket change to async because I can't spend too much time scrutinizing which of the `IQueryable<>` are legit EF queries and which are in fact `EnumerableQuery<>` wrappers created via `enumerable.AsQueryable<>` (believe me it's not straightforward in that code, and as a typical legacy system it has very little test coverage).

Since framework lets me down in this case, I end up creating global extension methods that shadow extension methods from `System.Data.Entity.QueryableExtensions` but can actually do async-style enumeration over a queryable that is backed by in-memory `IEnumerable<>`, which is far from ideal.

**Comment by roji** [comment]
>  Instead this should be default behavior of AsQueryable i.e. enumerable.AsQueryable().ToListAsync() should simply work, even if "async" enumeration means simply do it sync and pass to Task.FromResult under the hood. I don't understand why is this not the default behavior.

We will definitely not be doing something which implements async by doing sync under the hood - that is a violation of the contract. The method has "Async" in its name for a reason. I get that this could make sense specifically in the context of testing and mocking, but that's a very specific context, and it doesn't make sense to do that in a general operator that could be used elsewhere. You're free to do that yourself by implementing your own AsMyQueryable() - but I insist that this is going down the wrong road for mocking, with a lot of complexity where better options exist (see above). 

> I'm working with a legacy codebase that has a lot of blocking EF calls inherited from IEnumerable<> (e.g. ToList or ToArray) that should be changed to async equivalents. There's a lot of code and I want to simply blanket change to async because I can't spend too much time scrutinizing which of the IQueryable<> are legit EF queries and which are in fact EnumerableQuery<> wrappers created via enumerable.AsQueryable<>

You have production code which uses AsQueryable() and then composes queryable operators over that? Is there an actual reason for doing that? I'd recommend removing that, and making it very clear what's in-memory (by making it enumerable) and what's translated for remote execution (by leaving it queryable).

I understand you're looking for a quick way to "just make your legacy codebase async", and you can implement your own ToListAsync operators (or whatever) which work both over EF async queryables and over non-EF sync queryables, but that really is a hack that does not belong inside serious framework code in EF.

**Comment by DunetsNM** [comment]
> We will definitely not be doing something which implements async by doing sync under the hood - that is a violation of the contract. The method has "Async" in its name for a reason.

> that really is a hack that does not belong inside serious framework code in EF.

Really? Implementing async contract by simply lifting inherently synchronous code into a `Task<>` is not a hack but a very common practice found everywhere, including .NET core libs itself. For example:

https://github.com/dotnet/aspnetcore/blob/339c39e25bc95940f5103d978870f500463bd114/src/Identity/Extensions.Stores/src/UserStoreBase.cs#L591

Here `abstract class UserStoreBase` provides bunch of reasonable default method implementations that simply lift sync code into async interface instead of being made abstract or throwing an exception. 

Likewise, lifting inherently sync code such as enumeration of `EnumerableQuery<>` into a `Task` would be a reasonable default implementation of `ToListAsync()`.  

The only valid excuse to keep behavior as-is is backwards compatibility (I can only guess how much code out there really depends on current `NotSupportedException` behavior though). 

If `IQueryable<>` was designed from the scratch today it would make a lot more sense to inherit it from `IAsyncEnumerable<>` rather than `IEnumerable<>` and therefore ditch all blocking extension methods altogether.  But it'd still be fine to allow lifting `IEnumerable<>` into `IQueryable<>` via `.AsQueryable()` and then - yes - execute "async" extension methods on the result, with the suggested behavior.

**Comment by DunetsNM** [comment]
> You have production code which uses AsQueryable() and then composes queryable operators over that? Is there an actual reason for doing that? I'd recommend removing that, and making it very clear what's in-memory (by making it enumerable) and what's translated for remote execution (by leaving it queryable).

Those parts of the code lift in-memory `IEnumerable` such as `List<>` to `IQueryable<>` to apply some filtering and sorting via `Expression` values sourced from other places (not defending that design by any means).  Anyway, that was a small problem and it's fixed

I don't agree though that `IQueryable<>` always implies remote execution - only if it also implements `IDbAsyncEnumerable<>`.  Since `IQueryable` offers async extension methods for any queryables not just remote it should also do best effort to implement them for all, just like it does the opposite (supports non-sync extension methods both for remote and in-memory queryables).
Anyway I'm repeating myself.

**Comment by roji** [comment]
> Here abstract class UserStoreBase provides bunch of reasonable default method implementations that simply lift sync code into async interface instead of being made abstract or throwing an exception.

You're right that there are various cases where there's a default implementation of an async method, which calls into a sync counterpart; this is many times the case when an async API has been introduced into an abstraction (base class) which only had sync - to avoid breaking, you simply make the async delegate to sync. This is the case for System.Data for example, where e.g. DbConnection.OpenAsync() calls Open() by default.

So I agree I might have been overly harsh above on the proposal... I still believe that doing this makes sense for the above cases, e.g. backwards compat for an existing abstraction being retro-fitted with new async APIs - but should be generally avoided as much as possible. When users execute an async method, they should be reasonably confident that the method will not synchronously block, if at all possible. Now, if we could know that the thing to be executed synchronously doesn't actually block - because it's in-memory data - I think that would be acceptable (after all, async methods can cache and return synchronously when data is found in the cache). But we don't know that - the thing here could represent an actual synchronous database query, and so doing ToListAsync() for that case would really block the thread.

> The only valid excuse to keep behavior as-is is backwards compatibility (I can only guess how much code out there really depends on current NotSupportedException behavior though).

I don't think there's any backwards compat concern here - ToListAsync currently throws if the IQueryable it's given isn't an EF queryable (which can be executed asynchronously). What you're proposing (if I understood correctly) is to make it stop throwing, and to have it synchronously enumerate the results.

> I don't agree though that IQueryable<> always implies remote execution - only if it also implements IDbAsyncEnumerable<>.

I don't know what IDbAsyncEnumerable is - that doesn't exist in .NET. IQueryable indeed doesn't always (or by definition) imply remote execution, but that doesn't matter - the point is that it **can** imply remote execution.

To summarize, I do not agree that EF's ToListAsync() should synchronously enumerate the IQueryable it's given when that IQueryable can't be execute asynchronously; that would mean that in some (even many) cases, synchronous network access would be performed under the guise of an asynchronous method - that's something that's to be avoided. However, it really is trivial for you to simply write your own ToListAsync2() which does exactly this, and use that in your project; so do that if you feel that's right for your project.

**Comment by roji** [comment]
>> I don't know what IDbAsyncEnumerable is - that doesn't exist in .NET.
> 
> I meant this interface

That's the old, pre-core EF6 (note that this repo is for EF Core), just like the System.Data.Entity namespace which you mentioned above. EF6 is no longer being actively developed, and we'll definitely not be making any changes there.

> 00% totally agree with this. My reasoning is based on the assumption that if IQueryable can't execute asynchronously (doesn't implement IDbAsyncEnumerable or what's the latest test for it) then there's no network access / IO at all

Where does this assumption come from? There can be any number of enumerables in an application that perform network I/O, e.g. an enumerable coming out of Dapper, or any custom thing someone wrote that happens to perform I/O. If these are now wrapped in an IQueryable via AsQueryable(), you now have an IQueryable which doesn't support (EF's) async ToListAsync(), and at the same time performs sync I/O if enumerated synchronously. You may think that it's weird/contrived to wrap such an enumerable with AsQueryable(), but then you described above how AsQueryable() is being used to wrap in-memory enumerables in your legacy application - not just for test purposes; I find that just as weird...

I suspect you're approaching this with the narrow view of what's happening in your specific application; but since you're arguing for a behavior change for **all** applications in EF's ToListAsync, we need to take a much wider picture into account.

**Comment by DunetsNM** [comment]
> I suspect you're approaching this with the narrow view of what's happening in your specific application

My narrow app problem is long solved, and I agree that default async implementation should reduce to a possibility of sync I/O while implementing default async. If it's not strictly 0% (like in case of enumerating arbitrary IEnumerable) then it's safer to throw than "limp" / run potentially blocking I/O code and pretend that it's async.

However this is not what .NET always chooses to do (speaking of wider picture), there are multiple standards in default behavior of retrofitted async methods.  I can see see at least three patterns:

* `enumerable.AsQueryable.ToListAsync()` and its other async methods - throws NotSupportedException. 
* aforementioned `virtual DbConnection.OpenAsync` however does just that: runs sync Open by default and lies about being async. not even on the thread pool. 
* `TextReader.ReadLineAsync` and other async methods - invokes the sync counterpart on the thread pool.

In all those cases, sync methods existed long before Task-based async code was added to .NET.  When async methods were added they were brand new and not used by any pre-existing code, so framework architects had the freedom to just throw NotSupportedException by default too yet opted to limp by default rather than crash.

**Comment by roji** [comment]
> However this is not what .NET always chooses to do (speaking of wider picture), there are multiple standards in default behavior of retrofitted async methods. I can see see at least three patterns:

You're absolutely right. I'll just add that the examples where async methods call sync ones (e.g. DbConnection.OpenAsync) were introduced a long time ago - I'm not sure we'd do the same thing today. So it's not just that OpenAsync was retrofitted into an existing abstraction - and there was indeed some value in telling everyone "just use it, it'll always work, but not always be async" - it also just represented a different period in .NET itself, with slightly different practices and standards.

In any case, it seems we agree that failing-fast where we're not sure we won't block synchronously is generally better than "limping" (i.e. **maybe** doing the operation asynchronously, maybe blocking synchronously) - for new APIs, I don't think we should be bound by past decisions on unrelated APIs but rather do what we believe is right going forward.
